### PR TITLE
Added credential option

### PR DESCRIPTION
**Pull request by tonk**
As we use the TUS server with authentication (simple, plain `.htaccess`) I added `basic` authentication to the Tus client.

### CODE REVIEW

**Comment by adhocore** [comment]
thank you. please give me some time to have a look :)

**Comment by adhocore** [comment]
aha looks like we have merge conflict

**Comment by tonk** [comment]
Looks like we where both editing at the same time. I cloned this morning.
I also see the `-C` (nocolor) option I never saw before.

Do you want me to resolve the conflict, or do you do that?

**Comment by adhocore** [comment]
quick feedback:
should we instead have -U username and -P options?
-P doesn't accept value but will trigger a hidden prompt for password.
if -P is not given but -U is given with value then we use that value as a token that doesn't require password (and hence no prompt)

**Comment by adhocore** [comment]
i'm on mobile. so best if you go ahead ;)

**Comment by tonk** [comment]
Sry, can't have a prompt, should be able to run unattended. That why I have the `credentials` file.

**Comment by tonk** [comment]
I just resolved the conflicts (also made sure your `HEADER` var was kept) and left the authentication with the file.

Thank you very much.

File: creds [hunk 1]
```diff
@@ -0,0 +1,2 @@
+USER="test_user"
```
Comment (adhocore) [inline creds:1]: but we may not need to commit this file right?

Comment (tonk) [reply creds:1]: No, that's not needed, it's an example file

File: tusc.sh [hunk 2]
```diff
@@ -168,6 +176,7 @@ while [[ $# -gt 0 ]]; do
   case "$1" in
     -a | --algo) SUMALGO="$2"; shift 2 ;;
     -b | --base-path) BASEPATH="$2"; shift 2 ;;
+    -c | --CREDS) CREDS="$2"; shift 2 ;;
```
Comment (adhocore) [inline tusc.sh]: `-c | --creds` :)

Comment (tonk) [reply tusc.sh]: Yeah, that was stupid. Sorry.

**Comment by adhocore** [comment]
how about we keep it simple, addressing your need of unattended process?

we pass in `-c | --cred <ENV_VAR>` where $ENV_VAR holds the ready to use creds? this is for safety reasons that we don't define fixed var but upto the choice of user.

for example in .bashrc or .zshrc:
```
# it can be named anything as well
export MYTUSCRED=encoded_user_pass
```

then usage is `-c | --cred MYTUSCRED` (and not $MYTUSCRED)
tusc reads creds from the env var $MYTUSCRED.

and since $MYTUSCRED already holds well formatted and ready to use creds, we don't need to reformat it for use in $meta as done now. :)

File: creds (hunk 1 above)
Comment (adhocore) [reply creds:1]: hmm but we can document the structure of creds file.

anyway i have left another note which may change the way we pass creds.

**Comment by tonk** [comment]
Problem is that a password can be encoded, but it has to be supplied in plaintext to `curl`.
The user:pass is in plaintext on the `curl` commandline and the user in base64 in the metadata (so it ends up in TUS). So we need them both.

The reason I added the `meta` variable was to allow for unauthenticated as well as with user:pass combo. So you have the choice. 

If you want the `MYTUSCRED` it still needs the `--basic -u` flags. Not sure you gain something,

**Comment by adhocore** [comment]
i see. i will check and merge next day (will update few things as required) :)
thanks for using tusc and contributing. 🙂

**Comment by adhocore** [comment]
one more thing, i think for basic auth we pass in

base64(username: password) which is not plaintext, isn't the same for tus meta?

**Comment by tonk** [comment]
Yeah, you’re right, thought it was plaintext in url. Doesn’t enhance security to save as base64, as it can be decoded very easy

**Comment by adhocore** [comment]
hello, can we remove this part?
```
[[ x"${USER}" != x"" ]] && {
  USERB64="$(printf "%s" "${USER}" | base64)"
  PASSB64="$(printf "%s" "${PASS}" | base64)"
  TUSURL="$(echo "${TUSURL}" | sed "s!://!://${USERB64}:${PASSB64}@!")"
}
```

it exposes creds in url. in fact we dont need creds in url as curl request will contain creds as http header. what do you think?

**Comment by adhocore** [comment]
@tonk can i check with you if #6 works for you? :) thanks

**Comment by tonk** [comment]
Concerning the `TUSURL` thing, I'm not really sure if that works.
Problem is (I think, cannot verify as I'm not at the office) that the URL should hold the user and password and the `sed` inserts that.

**Comment by adhocore** [comment]
`curl --basic --user user:pass` already adds basic auth header required by the server :)

server can read that value in header: `Authorization: basic <base64(user:pass)>` :)

**Pull request merged**